docs/lazy-adr/adr-009-state-fraud-proofs.md

```diff
@@ -0,0 +1,233 @@
+
+```go
+
+// Data defines rollmint block data.
+type Data struct {
+	Txs                    Txs
+	IntermediateStateRoots IntermediateStateRoots
+	Evidence               EvidenceData
+}
+```
+
+These Intermediate State Roots (ISRs) are initially generated by a Rollmint sequencer during block execution which uses the ABCI interface.
+
+The following ABCI methods are called during block execution:
+`BeginBlock` at the start of a block
+`DeliverTx` for each transaction
+`EndBlock` at the end of a block
+
+After each of the above ABCI method calls, we generate an intermediate state root using a new ABCI method we introduce:
+
+```protobuf
+service ABCIApplication {
+    rpc GetAppHash(RequestGetAppHash) returns (ResponseGetAppHash);
+}
+
+message RequestGetAppHash {}
+
+message ResponseGetAppHash {
+  bytes app_hash = 1;
+}
+```
+
+This `GetAppHash` ABCI method returns an equivalent of `CommitID` hash in the ABCI method `Commit` and thus provides a way to extract ISRs from an app without doing any disk write operations.
+
+Full nodes use these ISRs to detect fraudulent state transitions. A full node must also execute all state transitions (`BeginBlock`, `DeliverTx`, and `EndBlock` calls) and compute its own Intermediate State Roots (ISRs). After each state transition, a full node compares the corresponding ISR with the ISR given by the Sequencer. If it finds a mismatch between its own computed ISR and and one given by the Sequencer, a fraudulent transition is detected and it moves on to generate a Fraud Proof.
+
+### Generating Fraud Proofs
+
+We introduce the following ABCI method to enable Fraud Proof Generation in the Cosmos SDK:
+
+```protobuf
+service ABCIApplication {
+    rpc GenerateFraudProof(RequestGenerateFraudProof) returns (ResponseGenerateFraudProof);
+}
+```
+
+With this new ABCI method, a Rollmint Full Node can send a request to a Cosmos SDK app to generate a Fraud Proof. In this request, it includes a list of all the state transitions from the start of the block upto the fraudulent state transition. The last non-nil state transition in this list corresponds to the fraudulent state transition.
+
+```protobuf
+message RequestGenerateFraudProof {
+  // All state transitions upto and including the fraudulent state transition
+  RequestBeginBlock begin_block_request = 1 [(gogoproto.nullable) = false];
+  repeated RequestDeliverTx deliver_tx_requests = 2;
+  RequestEndBlock end_block_request = 3;
+}
+```
+
+The `GenerateFraudProof` method in the Cosmos SDK app receives this list of state transitions and takes the following steps to generate a fraud proof:
+
+- Revert local state to the last committed state
+- Execute all the non-fraudulent state transitions
+- Remove all trace logs from executing the above state transitions
+- Execute the fraudulent state transition with tracing-enabled and store logs of what state, specifically key/value pairs, is accessed, modified, added, or removed during this fraudulent state transition.
```

Comment:
[non-blocking] We also discussed before that another option would be to leave tracing enabled always, right? So then you wouldn't need to re-execute all the state transitions to generate a fraud proof.

Reply:
I think this part will need to be updated in the future anyways, as our work-in-progress related to IAVL implementation details, shows that we need to change the direction slightly.